# 03 — Modelagem e Avaliação


In [34]:
%pip install -q imbalanced-learn
%pip install -q xgboost

from pathlib import Path
import pandas as pd
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_validate, cross_val_predict
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import RandomOverSampler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline as ImbPipeline
from sklearn.metrics import confusion_matrix

RANDOM_STATE = 42

RAW = Path("..") / "data" / "raw" / "df.csv"
PROCESSED = Path("..") / "data" / "processed"
TARGET = "STATUS"

pd.set_option("display.max_columns", None)

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [35]:
df = pd.read_csv(PROCESSED / "dataset_tratado.csv")

## 1. Split treino/teste

`stratify` preserva a proporção das classes nos dois conjuntos.


In [36]:
# Variáveis independentes
X = df.drop(columns=[TARGET])

# Variável Alvo
y = df[TARGET]

X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, stratify=y, random_state=RANDOM_STATE)

## 2. Modelos candidatos

Usar `Pipeline` evita vazamento: o scaler é ajustado só no fold de treino durante a validação cruzada.


In [37]:
# Cálculo do peso das classes para o XGBoost (Negativos / Positivos)
scale_pos_weight_val = (y_train == 0).sum() / (y_train == 1).sum()

modelos = {

  # XGBOOST (Nativo com peso de classe, sem SMOTE)
  "XGBoost": ImbPipeline([
    ("clf", XGBClassifier( scale_pos_weight=scale_pos_weight_val, random_state=RANDOM_STATE, eval_metric='logloss', n_jobs=-1))
  ]),

  # MODELOS BASEADOS EM DISTÂNCIA/GRADIENTE: COM Scaler + SMOTE
  "KNN": ImbPipeline([
    ("scaler", StandardScaler()),
    ("oversampler", RandomOverSampler(random_state=RANDOM_STATE)),
    ("clf", KNeighborsClassifier(n_neighbors=5))
  ]),
  "Regressão Logística": ImbPipeline([
    ("scaler", StandardScaler()),
    ("oversampler", RandomOverSampler(random_state=RANDOM_STATE)),
    ("clf", LogisticRegression(max_iter=1000, random_state=RANDOM_STATE))
  ]),
  "SVM": ImbPipeline([
    ("scaler", StandardScaler()),
    ("oversampler", RandomOverSampler(random_state=RANDOM_STATE)),
    ("clf", SVC(random_state=RANDOM_STATE))
  ]),

  # MODELOS BASEADOS EM ÁRVORES: SEM Scaler / SEM SMOTE
  "Floresta Aleatória": ImbPipeline([
    ("clf", RandomForestClassifier(n_estimators=100, random_state=RANDOM_STATE, n_jobs=-1, class_weight='balanced'))
  ]),
  "Árvore de Decisão": ImbPipeline([
    ("clf", DecisionTreeClassifier(random_state=RANDOM_STATE, class_weight='balanced'))
  ]),
}

## 3. Validação cruzada e Matriz Confusão


### Explicação teórica


|              | Previsto como **0**          | Previsto como **1**          |
| :----------- | :--------------------------- | :--------------------------- |
| **Real é 0** | **TN** (Verdadeiro Negativo) | **FP** (Falso Positivo)      |
| **Real é 1** | **FN** (Falso Negativo)      | **TP** (Verdadeiro Positivo) |

- **Posição `[0, 0]` — TN:** Quantidade de dados da classe **0** previstos como **0** (Acerto).
- **Posição `[0, 1]` — FP:** Quantidade de dados da classe **0** previstos como **1** (Erro / Alarme falso).
- **Posição `[1, 0]` — FN:** Quantidade de dados da classe **1** previstos como **0** (Erro / Deixou passar).
- **Posição `[1, 1]` — TP:** Quantidade de dados da classe **1** previstos como **1** (Acerto).

---

**Acurácia (_Accuracy_)**
Mede a taxa geral de acertos do modelo sobre todas as previsões realizadas.

Fórmula:
$$\text{Acurácia} = \frac{\text{TP} + \text{TN}}{\text{TP} + \text{TN} + \text{FP} + \text{FN}}$$

- **Resumo:** "De tudo que o modelo previu, quanto ele acertou?"
- **Quando usar:** Quando as classes do seu dataset estão **balanceadas** (ex: 50% classe 0 e 50% classe 1).
- **Armadilha:** Em dados desbalanceados (ex: 99% transações normais e 1% fraude), um modelo burro que prevê sempre "Normal" terá 99% de acurácia, mas será totalmente inútil porque não detectará nenhuma fraude.

---

**Precisão (_Precision_)**
Mede a exatidão das previsões da classe positiva.

Fórmula:
$$\text{Precisão} = \frac{\text{TP}}{\text{TP} + \text{FP}}$$

- **Resumo:** "Quando o modelo diz que é **Positivo (1)**, com que frequência ele está certo?"
- **Foco principal:** Minimizar **Falsos Positivos (FP)**.
- **Quando usar:** Quando o custo do alarme falso é muito alto (ex: filtro de SPAM — mover um e-mail importante do cliente/chefe para a caixa de SPAM por engano gera grande prejuízo).

---

**Recall / Sensibilidade (_Recall_)**
Mede a capacidade do modelo de encontrar todos os exemplos reais da classe positiva.

Fórmula:
$$\text{Recall} = \frac{\text{TP}}{\text{TP} + \text{FN}}$$

- **Resumo:** "De todos os casos **Positivos (1)** que existiam na realidade, quantos o modelo conseguiu capturar?"
- **Foco principal:** Minimizar **Falsos Negativos (FN)**.
- **Quando usar:** Quando o custo de "deixar passar" (omissão) é muito alto (ex: detecção de doenças graves ou detecção de fraudes financeiras).

---

**F1-Score**
É a média harmônica entre a **Precisão** e o **Recall**.

Fórmula:
$$\text{F1-Score} = 2 \times \frac{\text{Precisão} \times \text{Recall}}{\text{Precisão} + \text{Recall}}$$

- **Resumo:** "Qual é o equilíbrio geral entre Precisão e Recall?"
- **Quando usar:** Quando você trabalha com **datasets desbalanceados** e precisa equilibrar a redução de alarmes falsos (FP) sem deixar de capturar os casos reais (FN).
- **Vantagem:** Se a Precisão ou o Recall forem muito baixos, o F1-Score cai drasticamente, punindo modelos desequilibrados.

---

| Métrica      | Pergunta Principal                                     | Prioridade de Erro a Evitar        |
| :----------- | :----------------------------------------------------- | :--------------------------------- |
| **Acurácia** | De forma geral, quanto o modelo acertou?               | Erro geral (só usar se balanceado) |
| **Precisão** | Se disse que é 1, posso confiar?                       | Evita **Falsos Positivos (FP)**    |
| **Recall**   | O modelo conseguiu achar todos os 1s?                  | Evita **Falsos Negativos (FN)**    |
| **F1-Score** | Como está o equilíbrio entre ser preciso e abrangente? | Evita desequilíbrio entre FP e FN  |


### Análise


In [38]:
# Validação Cruzada Estratificada
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

# Métricas de avaliação
scoring = ['f1', 'accuracy', 'recall', 'precision']

resultados = []

matrizes_confusao = {}

for nome, modelo in modelos.items():

  # 1. Métricas da Validação Cruzada (sem leakage)
  scores = cross_validate(modelo, X_train, y_train, cv=cv, scoring=scoring, n_jobs=-1)

  # 2. Gera predições "out-of-fold" (sem leakage entre treino e validação)
  y_pred_cv = cross_val_predict(modelo, X_train, y_train, cv=cv, n_jobs=-1)

  # 3. Calcula a matriz de confusão acumulada dos 5 folds
  cm = confusion_matrix(y_train, y_pred_cv)
  cm_str = f"[[{cm[0][0]}, {cm[0][1]}], [{cm[1][0]}, {cm[1][1]}]]"

  # 4. Armazena os resultados
  resultados.append({
    "Modelo": nome,
    "F1-Score (Média)": scores['test_f1'].mean().round(2),
    "Accuracy (Média)": scores['test_accuracy'].mean().round(2),
    "Recall (Média)": scores['test_recall'].mean().round(2),
    "Precision (Média)": scores['test_precision'].mean().round(2),
    "Matriz Confusão [TN FP / FN TP]": cm_str
  })

# Cria um DataFrame com os resultados e ordena pelo Recall (Média) em ordem decrescente
df_ranking = pd.DataFrame(resultados).sort_values(by="Recall (Média)", ascending=False)
print(df_ranking.to_string(index=False))

             Modelo  F1-Score (Média)  Accuracy (Média)  Recall (Média)  Precision (Média) Matriz Confusão [TN FP / FN TP]
Regressão Logística              0.10              0.62            0.47               0.06      [[4716, 2747], [189, 165]]
                SVM              0.09              0.70            0.32               0.05      [[5344, 2119], [240, 114]]
            XGBoost              0.12              0.90            0.15               0.10        [[6977, 486], [301, 53]]
                KNN              0.06              0.84            0.11               0.04        [[6549, 914], [315, 39]]
  Árvore de Decisão              0.06              0.90            0.08               0.05        [[6977, 486], [326, 28]]
 Floresta Aleatória              0.06              0.93            0.06               0.08        [[7216, 247], [334, 20]]


## 4. Qual métrica foi levada em conta e por quê ?


**Objetivo:** Evitar perdas

**Métrica escolhida::** Recall

Em modelos de crédito, existem dois tipos de erros com custos:

Falso Negativo (FN): O modelo classifica o mau pagador como "bom", gerando perda direta por inadimplência (o valor total do calote que a empresa absorve).

Falso Positivo (FP): O modelo classifica o bom pagador como "mau", gerando um custo de oportunidade (deixar de ganhar a margem/juros de um bom cliente).

O custo de tomar um calote ($FN$) costuma ser dramaticamente superior ao custo de recusar um bom cliente ($FP$).

Dessa forma, como o objetivo da anállise é evitar perdas, a melhor métrica é o **Recall**, pois ele mede a capacidade do modelo de encontrar todos os exemplos reais da classe positiva, ou seja, os maus pagadores.


## 5. Comparação e Avaliação dos Modelos


**Análise dos Modelos**

Analisando a tabela de resultados.

| Modelo              | F1-Score | Accuracy | Recall | Precision | Avaliação no Contexto de Crédito                                                                                             |
| ------------------- | -------- | -------- | ------ | --------- | ---------------------------------------------------------------------------------------------------------------------------- |
| Regressão Logística | 0.10     | 0.62     | 0.47   | 0.06      | O "menos pior": Identifica a maior proporção de maus pagadores (Recall de 47%), embora a custa de muitos falsos alarmes. Captura a maior fatia de maus pagadores (47%, barrando 165 de 354 calotes). No entanto, assume um alto custo operacional ao gerar 2.747 falsos alarmes, derrubando a precisão para 6%.|
| SVM                 | 0.09     | 0.70     | 0.32   | 0.05      | Desempenho intermediário ineficiente: Pega cerca de um terço dos calotes (114 de 354, Recall de 32%), mas exige 2.119 falsos alarmes para isso, sem oferecer precisão competitiva (5%).|
|XGBoost              |0.12      |	0.90    |	0.15   |	0.10     |O mais equilibrado tecnicamente: Apresenta o maior F1-Score (0.12) e a maior Precisão (10%) entre os modelos que capturam risco. É seletivo: barra 53 calotes gerando apenas 486 alarmes falsos (preserva 90% de acurácia global).|
| KNN                 | 0.06     | 0.84     | 0.11   | 0.04      |Baixa eficiência operacional: Consegue barrar apenas 39 maus pagadores (Recall de 11%), mas gera 914 falsos alarmes, com precisão crítica de 4%.|
| Floresta Aleatória  | 0.06     | 0.93     | 0.06   | 0.08      |A armadilha da Acurácia: Apresenta 93% de acurácia apenas por aprovar quase todo mundo, mas falha na gestão de risco ao ignorar 94% dos calotes (deixa passar 334 de 354 maus pagadores).|
| Árvore de Decisão   | 0.06     | 0.90     | 0.08   | 0.05      | Fraco poder discriminatório: Ignora 92% dos maus pagadores (deixa passar 326 calotes reais) com apenas 5% de precisão. |

<br><br>
**Interpretação:**

Verifica-se um cenário muito claro de **forte desbalanceamento de classes** (muito mais bons pagadores do que maus pagadores).
<br><br>

**_Melhor modelo para ESTE caso:_**

Todos os modelos estão com desempenho muito fraco para um problema real de concessão de crédito.

Nesse cenário a _<u>Regressão Logística</u>_ apresentou o melhor comportamento prático.
<br><br>

**_Por quê a Regressão Logística?_**

Em análise de crédito, o custo de aprovar um mau pagador (inadimplência) é muito maior do que o custo de recusar um bom pagador.

Portanto, a métrica mais crítica no início é o Recall (a capacidade de encontrar os maus pagadores no meio da base).

A Regressão Logística conseguiu capturar 49% dos maus pagadores (Recall = 0.49).

A Floresta Aleatória, apesar de ter a maior acurácia (93%), só pegou 6% dos maus pagadores (Recall = 0.06). Ela basicamente aprendeu a dizer "Aprovado" para quase todo mundo para acertar a maioria dos casos.
<br><br>

**_Por que os resultados gerais estão baixos (F1 ~ 0.06 a 0.10)?_**

Uma precisão de ~0.05 e F1-Score tão baixo indicam que a classe de maus pagadores é extremamente minoritária.


## 5. Limitações

**Problemas encontrados**

- Há forte desbalanceamento de classes.

**Possíveis melhoras**

- Ajustar o Limiar de Decisão (Decision Threshold): baixar esse limiar para $10\%$ ou $20\%$ para tornar o modelo mais sensível à inadimplência.
- Muda o SVM para SVC(probability=True) para otimizar suas probabilidades.
- Implementar a estratégia de 3 Zonas de Decisão (Aprovação Direta, Mesa de Análise/Documentação e Recusa Automática), utilizando as probabilidades contínuas (predict_proba) geradas pelo modelo em vez de um corte seco em 50%.


In [ ]:
modelos = {
  
  # XGBOOST (Nativo com peso de classe, sem SMOTE)
  "XGBoost": ImbPipeline([
    ("clf", XGBClassifier( scale_pos_weight=scale_pos_weight_val, random_state=RANDOM_STATE, eval_metric='logloss', n_jobs=-1))
  ]),

 # MODELOS BASEADOS EM DISTÂNCIA/GRADIENTE: COM Scaler + SMOTE Moderado (20%)
  "Regressão Logística": ImbPipeline([
    ("scaler", StandardScaler()),
    ("oversampler", RandomOverSampler(sampling_strategy=PROPORCAO_OVERSAMPLING, random_state=RANDOM_STATE)),
    ("clf", LogisticRegression(max_iter=1000, random_state=RANDOM_STATE))
  ]),
  
  "SVM": ImbPipeline([
    ("scaler", StandardScaler()),
    ("oversampler", RandomOverSampler(sampling_strategy=PROPORCAO_OVERSAMPLING, random_state=RANDOM_STATE)),
    ("clf", SVC(probability=True, random_state=RANDOM_STATE))
  ]),

  "KNN": ImbPipeline([
    ("scaler", StandardScaler()),
    ("oversampler", RandomOverSampler(sampling_strategy=PROPORCAO_OVERSAMPLING, random_state=RANDOM_STATE)),
    ("clf", KNeighborsClassifier(n_neighbors=5))
  ]),

 # MODELOS BASEADOS EM ÁRVORES: SEM Scaler / SEM SMOTE (class_weight='balanced')
  "Floresta Aleatória": ImbPipeline([
    ("clf", RandomForestClassifier(n_estimators=100, random_state=RANDOM_STATE, n_jobs=-1, class_weight='balanced'))
  ]),
    
  "Árvore de Decisão": ImbPipeline([
    ("clf", DecisionTreeClassifier(random_state=RANDOM_STATE, class_weight='balanced'))
  ]),
}

resultados = []
matrizes_confusao = {}

for nome, modelo in modelos.items():
  scores = cross_validate(modelo, X_train, y_train, cv=cv, scoring=scoring, n_jobs=-1)
  
  y_pred_cv = cross_val_predict(modelo, X_train, y_train, cv=cv, n_jobs=-1)

  cm = confusion_matrix(y_train, y_pred_cv)
  cm_str = f"[[{cm[0][0]}, {cm[0][1]}], [{cm[1][0]}, {cm[1][1]}]]"

  resultados.append({
    "Modelo": nome,
    "F1-Score (Média)": scores['test_f1'].mean().round(2),
    "Accuracy (Média)": scores['test_accuracy'].mean().round(2),
    "Recall (Média)": scores['test_recall'].mean().round(2),
    "Precision (Média)": scores['test_precision'].mean().round(2),
    "Matriz Confusão [TN FP / FN TP]": cm_str
  })

df_ranking = pd.DataFrame(resultados).sort_values(by="Recall (Média)", ascending=False)
print(df_ranking.to_string(index=False))

             Modelo  F1-Score (Média)  Accuracy (Média)  Recall (Média)  Precision (Média) Matriz Confusão [TN FP / FN TP]
            XGBoost              0.12              0.90            0.15               0.10        [[6977, 486], [301, 53]]
                KNN              0.06              0.86            0.09               0.04        [[6692, 771], [321, 33]]
  Árvore de Decisão              0.06              0.90            0.08               0.05        [[6977, 486], [326, 28]]
 Floresta Aleatória              0.06              0.93            0.06               0.08        [[7216, 247], [334, 20]]
Regressão Logística              0.07              0.96            0.04               0.78          [[7456, 7], [341, 13]]
                SVM              0.07              0.95            0.04               0.25         [[7420, 43], [340, 14]]
